### PR TITLE
Eagerly instantiate closure/coroutine-like bounds with placeholders to deal with binders correctly

### PR DESCRIPTION
**Pull request by compiler-errors**
A follow-up to #119849, however it aims to fix a different set of issues. Currently, we have trouble confirming goals where built-in closure/fnptr/coroutine signatures are compared against higher-ranked goals.

Currently, we don't support goals like `for<'a> fn(&'a ()): Fn(&'a ())` because we don't expect the self type of goal to reference any bound regions from the goal, because we don't really know how to deal with the double binder of predicate + self type. However, this definitely can be reached (#121653) -- and in fact, it results in post-mono errors in the case of #112347 where the builtin type (e.g. a coroutine) is hidden behind a TAIT.

The proper fix here is to instantiate the goal before trying to extract the signature from the self type. See final two commits.

r? lcnr

### CODE REVIEW

File: tests/ui/higher-ranked/builtin-closure-like-bounds.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
+//@ edition:2024
+//@ compile-flags: -Zunstable-options
+//@ revisions: current next
+//@[next] compile-flags: -Znext-solver
+//@ check-pass
+
```
Comment (lcnr) [inline tests/ui/higher-ranked/builtin-closure-like-bounds.rs:6]: please add a comment here explaining this test

**Comment by lcnr** [comment]
this prevents an codegen ICE. Please add it as a test
```rust
trait Trait {
    type Assoc<'a>: FnOnce(&'a ());
}

impl Trait for () {
    type Assoc<'a> = fn(&'a ());
}

trait Indir {
    fn break_me() {}
}

impl<F: Trait> Indir for F
where
    for<'a> F::Assoc<'a>: FnOnce(&'a ()),
{
    fn break_me() {}
}

fn foo<F: Trait>() {
    F::break_me()
}

fn main() {
    foo::<()>();
}
```

@rust-lang/types we previously failed to handle builtin `Fn`-trait (and related traits) impls if the self type referenced a bound variable. We simply ended up considering such goals to not hold, which is incomplete. This mostly affects manually written where-bounds such as `for<'a> fn(&'a ()): FnOnce(&'a ())`. It also resulted in codegen ICE in somewhat artificial cases. This constraint unnecessary, as we can simply avoid it by instantiating higher ranked goals more eagerly.

This PR allows more code to pass and can in likely very artificial cases be breaking (by fixing coherence/changing candidate selection). I am confident enough that this won't break anything that I am going to start the FCP right away, even before we've finished the crater run. We can always reconsider if any breakage is found

@rfcbot fcp merge

**Comment by compiler-errors** [comment]
@craterbot check

**Comment by compiler-errors** [comment]
Crater run is clean.

One failure is `lu_packets`, which is known bad (it has impl-sorting dependent overflow errors). Others are disk out of space, except for [`wpilog-rs`](https://crater-reports.s3.amazonaws.com/pr-122267/try%2305804f728f276cf7ecedbf71c814df440d37f293/gh/oh-yes-0-fps.wpilog-rs/log.txt) which has no apparent failure reason and builds fine locally.

File: tests/ui/higher-ranked/closure-bound-codegen-ice.rs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+//@ revisions: current next
+//@[next] compile-flags: -Znext-solver
+//@ build-pass
+
+trait Trait {
```
Comment (lcnr) [inline tests/ui/higher-ranked/closure-bound-codegen-ice.rs:5]:
```suggestion
// Regression test for incomplete handling of Fn-trait goals,
// fixed in #122267.

trait Trait {
```

**Review by lcnr — COMMENTED** [review]
r=me after adding comment

**Pull request merged**